crab/actions/visual_prompt_actions.py

```diff
@@ -0,0 +1,282 @@
+# =========== Copyright 2024 @ CAMEL-AI.org. All Rights Reserved. ===========
+# Licensed under the Apache License, Version 2.0 (the “License”);
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an “AS IS” BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =========== Copyright 2024 @ CAMEL-AI.org. All Rights Reserved. ===========
+from functools import cache
+
+import torch
+from PIL import Image, ImageDraw, ImageFont
+from transformers import AutoProcessor, GroundingDinoForObjectDetection
+
+from crab import action
+from crab.utils.common import base64_to_image, image_to_base64
+
+try:
+    import easyocr
+    import numpy as np
+    import torch
+
+    device = "cuda" if torch.cuda.is_available() else "cpu"
+
+except ImportError:
+    pass
+
+BoxType = tuple[int, int, int, int]
+
+
+def calculate_iou(box1, box2):
+    xA = max(box1[0], box2[0])
+    yA = max(box1[1], box2[1])
+    xB = min(box1[2], box2[2])
+    yB = min(box1[3], box2[3])
+
+    interArea = max(0, xB - xA) * max(0, yB - yA)
+    box1Area = (box1[2] - box1[0]) * (box1[3] - box1[1])
+    box2Area = (box2[2] - box2[0]) * (box2[3] - box2[1])
+    unionArea = box1Area + box2Area - interArea
+    iou = interArea / unionArea
```

Comment:
unionArea could be 0 when one of the box area is zero or there is no intersection area between two boxes
`if unionArea == 0:
    return 0
`

Reply:
The union area is the whole occupied area of of box1 and box2. So I think `UnionArea==0` only happens when box1 and box2 are both 0, which may be impossible in our situation. But checking can be useful some time in the future. Thanks